distributed/tests/test_cancelled_state.py

```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
         a.block_get_data = False
         # Everything should still be executing as usual after this
         assert await c.submit(sum, c.map(inc, range(10))) == sum(map(inc, range(10)))
+
+
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_in_flight_lost_after_resumed(c, s, b):
+    block_get_data = asyncio.Lock()
+    in_get_data = asyncio.Event()
+
+    executing = Event()
+    lock_executing = Lock()
+
+    def block_execution(x, event, lock):
+        event.set()
+        with lock:
+            return x
+
+    class BlockedGetData(Worker):
+        async def get_data(self, comm, *args, **kwargs):
+            in_get_data.set()
+            async with block_get_data:
+                return await super().get_data(comm, *args, **kwargs)
+
+    async with BlockedGetData(s.address, name="blocked-get-dataworker") as a:
+        fut1 = c.submit(
+            block_execution,
+            1,
+            executing,
+            lock_executing,
+            workers=[a.address],
+            allow_other_workers=True,
+        )
+        await fut1
+        await lock_executing.acquire()
+        in_get_data.clear()
+        await block_get_data.acquire()
+        fut2 = c.submit(inc, fut1, workers=[b.address])
+
+        await in_get_data.wait()
+        await s.remove_worker(a.address, "foo", close=False, safe=True)
+
+        while not b.tasks[fut1.key].state == "resumed":
+            await asyncio.sleep(0.1)
```

Comment:
Am I missing a reason why fut1 is guaranteed to already b in b here?
```suggestion
        assert fut1.key in b.tasks
        while b.tasks[fut1.key].state != "resumed":
            await asyncio.sleep(0.01)
```